lib/spack/spack/dag.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import spack
+import spack.binary_distribution as bindist
+import spack.ci
+import spack.cmd
+import spack.main
+import spack.mirror
+import spack.paths
+import spack.repo
+
+
+def generate_snakefile(env, print_summary, output_file):
+
+    with spack.concretize.disable_compiler_existence_check():
+        with env.write_transaction():
+            env.concretize()
+            env.write()
+
+    phases = []
+    phases.append({
+        'name': 'specs',
+        'strip-compilers': False,
+    })
+
+    # Speed up staging by first fetching binary indices from all mirrors
+    # (including the per-PR mirror we may have just added above).
+    bindist.binary_index.update()
+
+    staged_phases = {}
+    for phase in phases:
+        phase_name = phase['name']
+        with spack.concretize.disable_compiler_existence_check():
+            staged_phases[phase_name] = spack.ci.stage_spec_jobs(
+                env.spec_lists[phase_name],
+                check_index_only=True)
+
+    # There should only be one phase for now, so one snakefile
+    for phase in phases:
+        phase_name = phase['name']
+
+        # We care about dependencies
+        spec_labels, dependencies, stages = staged_phases[phase_name]
```

Comment:
hey @scottwittenburg ! I know you have a lot on your plate so I was thinking maybe I could help with this? If you have a few minutes to async discuss (even just in this issue!) your idea/vision for how to refactor the data structures - that could be enough to get me going, and then you could just give guidance as we work! I think @alecbcs is interested in this too.

Reply:
Hi @v, yes, I hoped I could tackle this quickly, but I haven't got to it yet.  Sorry to leave it up in the air like that, but thanks for being patient 😁.  If you're interested in giving it a go, that would be awesome.  Would it work better for you to chat about it, or for me to write something up, describing how it works, pointing at certain bits of the code, and discussing how we could change it?  

There are some upcoming changes (e.g. moving to using the full hash instead of dag hash to describe everything) that will complicate the situation a little, but I'm sure we can handle that.  Some of the changes we'll want to make will require investigation/experimentation to see if actual spack behaves like we imagine/hope it might.  

If you're interested to set up a call for a design meeting, maybe we can include @alecbcs.  If you prefer me to prepare something written you can digest offline, that's fine too, I could create an issue where we could iterate on the design.

Reply:
Both of those options will be great! I do think @alecbcs wants to be included, so I'll catch up with him to see when/what he could do (he's in classes right now). Will get back to you after we chat!